Adding svelte-query

AdityaHegde opened the pull request.

ericpgreen2 commented:
Here's some suggestions for naming – so we can start aligning with the `MetricView` entity type and avoid having the `RillRuntime` prefix in too many places:
- Rename file `RillRuntimeMetricsController.ts` to `MetricViewsController.ts`
- Rename file `RillRuntimeMetricsActions.ts` to `MetricViewsActions.ts`
- Rename action `getRuntimeTimeSeries` to `getMetricViewTimeseries`
- Rename action `getRuntimeMetricsMeta` to `getMetricViewMeta`
- Rename action `getRuntimeTopList` to `getMetricViewToplist`
- Rename type `RuntimeMetricsMetaResponse` to `MetricViewMetaResponse`

I think we can remove this old file:
- Delete `MetricsExplorerActions.ts`

I ran into these errors in `RillRuntimeMetricsActions.ts`:
- Property 'getManyByField' does not exist on type 'MeasureDefinitionStateService'.
- Property 'getRawResponse' does not exist on type 'typeof ActionResponseFactory'.
- TypeError: Cannot read properties of undefined (reading 'metricDefLabel') at RillRuntimeMetricsActions.getRuntimeMetricsMeta (/Users/ericgreen/Desktop/Rill/Code/rill-developer/src/common/rill-developer-service/RillRuntimeMetricsActions.ts:111:39)

Let's use  `api/v1/metric-views/{view-name}/` as a prefix for each of these APIs

ericpgreen2 commented:
And as mentioned, let's split out this backend PR from each of the frontend PRs:
- one for `/meta`: [svelte-query-meta](https://github.com/rilldata/rill-developer/tree/svelte-query-meta)
- one for `/timeseries`: [svelte-query-timeseries](https://github.com/rilldata/rill-developer/tree/svelte-query-timeseries)
- one for `/toplist`
- one for `/big-number`

AdityaHegde commented:
Updated the names.

Splitting this PR would duplicated the changes to add the endpoint. The only this specific are the methods in Actions class. I feel we should have a PR with outline of the backend here and the UI changes will come in later PRs.

We cannot remove MetricsExplorerActions until everything has been moved to use the new endpoints. This will come after the final PR has been merged.

I have added api/v1. But i feel using view name instead of ID doesnt make sense. We can change the minute details in later PRs.

The errors you got were probably because of cached type definitions. Open the classes where they are defined and it will get fixed. They work without issues for me and also compiles without issues.

ericpgreen2 commented:
Regarding the PR split, that's what I meant – keep this as the backend PR, and do the frontend work in different PRs.

Using `metricViewId` not `metricViewName` sounds good to me.

ericpgreen2 commented:
Remaining TODO list here:
- [ ] Eliminate line chart jank
- [ ] Add `/totals` API and query it via `svelte-query`
- [ ] Fix CI/CD errors

ericpgreen2 commented:
Here are the last things IMO:
- #823 
- #825 
- To eliminate the Explorer Redux store entirely, we'd remove these functions: `bootstrapMetricsExplorer`, `syncExplorer`, `syncDimensions`, `syncMeasures`. Can we do that? "Syncing" shouldn't be necessary – we should use the `svelte-query` model of invalidating the cache (when necessary) and `svelte-query` will autonomously refetch/sync in the background.

And one bit of work that I think we should punt on is making the `svelte-query` component code more concise. I think we _could_ move from code that looks like this...
```
let queryKey = getMetricViewMetaQueryKey(metricsDefId);
const queryResult = useQuery<MetricViewMetaResponse, Error>(queryKey, () =>
  getMetricViewMetadata(metricsDefId)
);
$: {
  queryKey = getMetricViewMetaQueryKey(metricsDefId);
  queryResult.setOptions(queryKey, () => getMetricViewMetadata(metricsDefId));
}
```
... to something like this ...
```
$: metaQueryResult = useMetaQuery({metricsDefId });
```
... but I think we should just wait until we add OpenAPI and a client code generator (like https://orval.dev/guides/svelte-query). The generated code will include those concise custom hooks, so let's just keep this as our motivation to switch to OpenAPI & code generation.

ericpgreen2 commented:
TODOs:
- [ ] Fix linting errors
- [ ] Humanise the dimension values
- [ ] Some leaderboards get stuck in a loading state (see nyc311 dataset)
- [ ] Switching between dashboards sometimes shows the wrong time grain

hamilton commented on src/routes/_surfaces/workspace/explore/leaderboards/LeaderboardDisplay.svelte:
```diff
@@ -1,105 +1,109 @@
   let referenceValue: number;
 
-  $: if ($bigNumberEntity && $measureField) {
+  $: totalsQuery = useGetMetricViewTotals(
```
why is `totalsQuery` reactive while `referenceValueQuery` is not? This is a confusing pattern.

ericpgreen2 replied:
I think `referenceValueQuery` is supposed to be reactive too. @AdityaHegde touched that so I'll just let him confirm.

AdityaHegde replied:
Fixed!

hamilton commented on src/routes/_surfaces/workspace/explore/leaderboards/LeaderboardDisplay.svelte:
```diff
@@ -1,105 +1,109 @@
+    }
+  );
+
+  $: if ($totalsQuery && $referenceValueQuery && activeMeasure?.sqlName) {
```
This doesn't seem to work since the leaderboard values aren't updating

hamilton replied:
This statement should probably be re-written. In each case, you can check something like this:

```javascript
    if (activeMeasureName in $totalsQuery.data.data) {
      referenceValue = $totalsQuery.data.data[activeMeasureName];
    }
```

And it should work.

AdityaHegde replied:
`in` check is not needed, just checking for existence of response is enough.
